File: Kotlin/src/main/kotlin/programmers/devMatching2021/Backend_ex1.kt
package programmers.devMatching2021

private class Solution {
    fun solution(lottos: IntArray, win_nums: IntArray): IntArray {
        val n = lottos.size
        var cor = 0
        var zero = 0
        var wrong = 0

        return lottos.map {
            when (it) {
                0 -> zero++
                in win_nums -> cor++
                else -> wrong++
            }
        }.run { when {
            cor == n -> intArrayOf(1, 1)
            zero == n -> intArrayOf(1, 6)
            wrong == n -> intArrayOf(6, 6)
            else -> intArrayOf(wrong + 1, n - cor + 1)
        }}
    }
}

private class Solution2 {
    fun solution(lottos: IntArray, win_nums: IntArray): IntArray {
        return intArrayOf(
            lottos.size.plus(1) - lottos.filter { it in win_nums || it == 0 }.size,
            lottos.size.plus(1) - lottos.filter(win_nums::contains).size
        )
            .map { if (it > 6) it - 1 else it }
            .toIntArray()
    }
}

fun backend_ex1() {
    val lottos = intArrayOf(0, 0, 0, 0, 0, 0)
    val win_nums = intArrayOf(38, 19, 20, 40, 15, 25)
    val (max, min) = Solution().solution(lottos, win_nums)
    println("$max $min")
}

File: Cpp/Programmers/devMatching2021/Backend_ex1.cpp
#include <bits/stdc++.h>
#define fastio ios::sync_with_stdio(0)->cin.tie(0)

using namespace std;

static vector<int> solution(vector<int> lottos, vector<int> win_nums) {
    int zero = 0;
    int cor = 0;
    int wrong = 0;
    int n = win_nums.size();

    for (int lotto : lottos) {
        if (lotto == 0) zero++;
        else if (std::find(win_nums.begin(), win_nums.end(), lotto) != win_nums.end()) cor++;
        else wrong++;
    }

    vector<int> ans;

    if (cor == n) ans = { 1, 1 };
    else if (wrong == n) ans = { 6, 6 };
    else if (zero == n) ans = { 1, 6 };
    else ans = { wrong + 1, wrong + zero + 1 };

    return ans;
}


void backend_ex1() {
    vector<int> lottos = { 45, 4, 35, 20, 3, 9 };
    vector<int> win_nums = { 20, 9, 3, 45, 4, 35 };
    vector<int> ans = solution(lottos, win_nums);
    for (int a : ans) cout << a << " ";
}